File: maquette/charte.js
// Réglages de design des maquettes : le brouillon de scrutin/charte.py.
//
// `CHARTE` porte les valeurs par défaut (palette validée, PLAN_MODERNISATION.md
// Partie 7). Chaque variante passe sa propre surcharge à `appliquerCharte` :
// c'est là que se comparent les partis pris graphiques.
//
// On ne touche jamais aux données des figures — figures.js est généré.
window.CHARTE = {
  surface: "#fcfcfb",     // fond de la figure
  encre1: "#1f1f1c",      // texte principal
  encre2: "#5f5e58",      // texte secondaire, axes
  grille: "#e1e0d9",      // grille hairline
  bleu: "#2a78d6",        // oui / confirmé
  bleuClair: "#86b6ef",   // extrapolé (même teinte, plus clair = estimé)
  rouge: "#c9352b",       // non
  neutre: "#f0efec",      // milieu de l'échelle divergente, ancré à 50 %
  fonte: "system-ui, -apple-system, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif",
  carte: {
    demiEtendue: 20,      // bornes symétriques autour de 50 : 30 % … 70 %
    zoom: null,           // Mapbox seulement : null = garder le zoom de la figure
    opacite: 0.95,
    contour: 0.2,         // filet entre communes, dans la couleur du fond
    hauteur: 420,
    barreCouleur: true,
  },
  histogramme: {
    hauteur: 360,
    ligneMajorite: true,
  },
};

window.CONFIG_PLOTLY = { displayModeBar: false, responsive: true, displaylogo: false };

// Applique la charte par-dessus une figure exportée par construire.py.
// `genre` : "histogramme", "carte" (Mapbox) ou "carte_svg" (projection SVG).
// `surcharge` : les valeurs propres à la variante.
window.appliquerCharte = function (fig, genre, surcharge) {
  const s = surcharge || {};
  const c = Object.assign({}, window.CHARTE, s);
  c.carte = Object.assign({}, window.CHARTE.carte, s.carte);
  c.histogramme = Object.assign({}, window.CHARTE.histogramme, s.histogramme);

  const data = JSON.parse(JSON.stringify(fig.data));
  const layout = JSON.parse(JSON.stringify(fig.layout));

  layout.font = { family: c.fonte, color: c.encre1, size: c.taillePolice || 14 };
  layout.paper_bgcolor = c.surface;
  layout.plot_bgcolor = c.surface;
  layout.margin = { l: 8, r: 8, t: 8, b: 8 };

  if (genre === "histogramme") {
    // Deux traces : « Déja dépouillés », puis « Extrapolés ».
    const couleurs = c.couleursSeries || [c.bleu, c.bleuClair];
    data.forEach((trace, i) => {
      trace.marker = Object.assign({}, trace.marker, { color: couleurs[i] });
      trace.textposition = "outside";
      trace.textfont = { color: c.encre2 };
    });
    layout.height = c.histogramme.hauteur;
    layout.margin = { l: 48, r: 8, t: 16, b: 40 };
    layout.legend = { orientation: "h", y: -0.18, title: { text: "" } };
    layout.xaxis = Object.assign({}, layout.xaxis, { title: { text: "" }, showgrid: false });
    layout.yaxis = Object.assign({}, layout.yaxis, {
      title: { text: "" }, tickformat: ".0%", gridcolor: c.grille, zeroline: false, range: [0, 1],
    });
    if (c.histogramme.ligneMajorite) {
      // L'information n°1 d'une votation : le seuil qu'il faut franchir.
      layout.shapes = [{
        type: "line", xref: "paper", x0: 0, x1: 1, y0: 0.5, y1: 0.5,
        line: { color: c.encre2, width: 1, dash: "dot" },
      }];
      layout.annotations = [{
        xref: "paper", x: 1, y: 0.5, xanchor: "right", yanchor: "bottom",
        text: "majorité", showarrow: false, font: { size: 12, color: c.encre2 },
      }];
    }
  }

  if (genre === "carte" || genre === "carte_svg") {
    // Les contours sont stockés une fois pour toutes (voir construire.py).
    data.forEach((trace) => { if (!trace.geojson) trace.geojson = window.GEOJSON; });

    const d = c.carte.demiEtendue;
    layout.coloraxis = Object.assign({}, layout.coloraxis, {
      // Divergente ancrée à 50 % : « penche non » ← neutre → « penche oui ».
      colorscale: [[0, c.rouge], [0.5, c.neutre], [1, c.bleu]],
      cmin: 50 - d, cmid: 50, cmax: 50 + d,
      showscale: c.carte.barreCouleur,
      colorbar: {
        title: { text: "" }, ticksuffix: " %", thickness: 8, len: 0.7, x: 1,
        outlinewidth: 0, tickfont: { color: c.encre2, size: 11 },
      },
    });
    data.forEach((trace) => {
      trace.marker = Object.assign({}, trace.marker, {
        opacity: c.carte.opacite,
        line: { width: c.carte.contour, color: c.surface },
      });
    });
    layout.height = c.carte.hauteur;
    layout.margin = { l: 0, r: 0, t: 0, b: 0 };
    if (genre === "carte_svg") {
      layout.geo = Object.assign({}, layout.geo, {
        bgcolor: "rgba(0,0,0,0)", showframe: false, showcoastlines: false,
      });
    } else {
      layout.mapbox = Object.assign({}, layout.mapbox, {
        // Pas de tuiles externes : un simple aplat, à la couleur de la page.
        style: { version: 8, sources: {}, layers: [{ id: "fond", type: "background", paint: { "background-color": c.surface } }] },
      });
      // Contrairement à la projection SVG (`fitbounds`), Mapbox garde le zoom
      // figé à la construction : c'est à la page de le recalculer.
      if (c.carte.zoom) layout.mapbox.zoom = c.carte.zoom;
      if (c.carte.centre) layout.mapbox.center = c.carte.centre;
    }
  }

  return { data, layout };
};

// Raccourci : dessine une figure de figures.js dans un conteneur.
window.tracer = function (cible, fig, genre, surcharge) {
  const p = window.appliquerCharte(fig, genre, surcharge);
  return Plotly.newPlot(cible, p.data, p.layout, window.CONFIG_PLOTLY);
};

File: scrutin/charte.py
"""Palette et réglages Plotly, repris de la maquette D′ (couleurs aussi dans style.css)."""

from django.templatetags.static import static

ENCRE = "#1f1f1c"
SURFACE = "#ffffff"
BLEU = "#2a78d6"     # oui
ROUGE = "#c9352b"    # non
NEUTRE = "#f0efec"   # 50 %
GRIS = "#5f5e58"     # texte secondaire (--gris)
POINT = "#9d9c96"    # point au repos dans un nuage (--fleche)
GRILLE = "#e1e0d9"   # filets des axes (--bord)
FONTE = "system-ui, -apple-system, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif"

LACS = "carte/lacs.geojson"

DEMI_ETENDUE = 18    # échelle de la carte : 50 ± 18 %
CONTOUR = 0.15       # filet entre communes, en px

CONFIG_CARTE = {
    "displayModeBar": "hover",
    "modeBarButtons": [["zoomInMap", "zoomOutMap", "resetViewMap"]],
    "responsive": True,
    "displaylogo": False,
    "scrollZoom": True,
}


CONFIG_NUAGE = {
    "displayModeBar": "hover",
    "modeBarButtons": [["zoom2d", "pan2d", "resetScale2d"]],
    "responsive": True,
    "displaylogo": False,
    # Pas de zoom à la molette : sur un nuage, il vole le défilement de la page.
    "scrollZoom": False,
}


def habiller_carte(figure, emprise, echelle=None):
    """Couleurs, fond sans tuiles et limites de déplacement.

    ``echelle`` : (min, milieu, max) de l'axe de couleur, par défaut le % de oui.
    L'emprise part dans ``layout.meta`` : ``cartes.js`` en tire centre et zoom.
    """
    (x0, y0), (x1, y1) = emprise
    cmin, cmid, cmax = echelle or (50 - DEMI_ETENDUE, 50, 50 + DEMI_ETENDUE)
    largeur, hauteur = x1 - x0, y1 - y0
    figure.update_layout(
        font={"family": FONTE, "color": ENCRE},
        paper_bgcolor=SURFACE,
        margin={"l": 0, "r": 0, "t": 0, "b": 0},
        meta={"emprise": [[x0, y0], [x1, y1]]},
        coloraxis={
            "colorscale": [[0, ROUGE], [0.5, NEUTRE], [1, BLEU]],
            "cmin": cmin, "cmid": cmid, "cmax": cmax,
            "showscale": False,
        },
        map={
            "style": {"version": 8, "sources": {}, "layers": [
                {"id": "fond", "type": "background", "paint": {"background-color": SURFACE}},
            ]},
            # Les lacs par-dessus les communes : dans swissBOUNDARIES3D, une
            # commune riveraine s'étend jusqu'au milieu de l'eau.
            "layers": [{"sourcetype": "geojson", "source": static(LACS),
                        "type": "fill", "color": SURFACE}],
            "center": {"lon": (x0 + x1) / 2, "lat": (y0 + y1) / 2},
            "zoom": 6,
            # On ne s'éloigne pas du pays de plus de sa demi-largeur.
            "bounds": {"west": x0 - largeur / 2, "east": x1 + largeur / 2,
                       "south": y0 - hauteur / 2, "north": y1 + hauteur / 2},
        },
    )
    figure.update_traces(marker_opacity=1, marker_line_width=CONTOUR, marker_line_color=SURFACE)
    return figure


def habiller_nuage(figure, titre_x, titre_y):
    """Fond, grille et zéro des deux nuages ACP. Le zéro est plus marqué que la
    grille : sur une ACP, c'est le signe d'une coordonnée qui porte le sens."""
    axe = {
        "showgrid": True, "gridcolor": GRILLE, "gridwidth": 1,
        "zeroline": True, "zerolinecolor": GRIS, "zerolinewidth": 1,
        "showline": False, "ticks": "outside", "tickcolor": GRILLE,
        "title": {"font": {"size": 13}},
    }
    figure.update_layout(
        font={"family": FONTE, "color": ENCRE, "size": 12},
        paper_bgcolor=SURFACE,
        plot_bgcolor=SURFACE,
        margin={"l": 52, "r": 16, "t": 10, "b": 44},
        showlegend=False,
        hovermode="closest",
        xaxis={**axe, "title": {**axe["title"], "text": titre_x}},
        yaxis={**axe, "title": {**axe["title"], "text": titre_y}},
    )
    return figure


def demi_etendue(valeurs):
    """Demi-étendue d'une échelle de couleur centrée sur zéro : le 95ᵉ centile des
    écarts, pour qu'une poignée de valeurs extrêmes n'écrase pas les nuances."""
    ecarts = sorted(abs(v) for v in valeurs)
    return round(ecarts[int(0.95 * (len(ecarts) - 1))], 2) or 1.0
